### PR TITLE
Non-animated property change notifications

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As described in #3109, transitions were not working correctly because there was no way to get change notifications for non-animated property changes.

This PR is the first part in fixing #3109 - it does NOT fix transitions, but it does provide a way to get property change notifications for non-animated property changes. Transitions are fixed in #3854 

- Pass information for all property changes to `OnPropertyChangedCore`; whether they result in an effective value change or not
- `AvaloniaPropertyChangedEventArgs` now has an `IsEffectiveValue` property which will be false when a higher priority value is active
- `OnPropertyChangedCore` calls `OnPropertyChanged` with all notifications which have `IsEffectiveValue == true`
- Added `AvaloniaObject.GetBaseValue` to get a value with a specified priority
- Change the signature of `OnPropertyChanged` again (sorry) to take an `AvaloniaPropertyChangedEventArgs<T>`

## Checklist

- [x] Added unit tests (if possible)?

## Breaking changes

The signature for `OnPropertyChanged` has changed again.

### CODE REVIEW

**Comment by MarchingCube** [comment]
In case somebody wonders about perf impact:

This PR
|         Method |        Mean |       Error |      StdDev |    Gen 0 |    Gen 1 |   Gen 2 |  Allocated |
|--------------- |------------:|------------:|------------:|---------:|---------:|--------:|-----------:|
| CreateCalendar | 20,164.0 us | 224.2566 us | 209.7698 us | 687.5000 | 281.2500 | 31.2500 | 3962.71 KB |
|   CreateButton |    139.0 us |   0.9656 us |   0.8063 us |   6.8359 |   0.9766 |       - |   28.66 KB |
|  CreateTextBox |  1,080.0 us |   4.4062 us |   3.6794 us |  42.9688 |  21.4844 |       - |  183.04 KB |

Master
|         Method |        Mean |      Error |     StdDev |    Gen 0 |    Gen 1 | Gen 2 |  Allocated |
|--------------- |------------:|-----------:|-----------:|---------:|---------:|------:|-----------:|
| CreateCalendar | 19,973.4 us | 232.356 us | 194.027 us | 593.7500 | 250.0000 |     - | 3630.23 KB |
|   CreateButton |    140.4 us |   2.530 us |   2.366 us |   6.3477 |   0.9766 |     - |   26.77 KB |
|  CreateTextBox |  1,114.1 us |  19.217 us |  17.976 us |  41.0156 |  19.5313 |     - |  172.22 KB |

So perf is withing margin of error and we now need roughly 10% more memory.

File: src/Avalonia.Base/IAvaloniaObject.cs [hunk 1]
```diff
@@ -41,6 +41,19 @@ public interface IAvaloniaObject
...
+        /// </summary>
+        /// <typeparam name="T">The type of the property.</typeparam>
+        /// <param name="property">The property.</param>
+        /// <param name="maxPriority">The maximum priority for the value.</param>
+        /// <remarks>
+        /// Gets the value of the property, if set on this object with a priority equal or lower to
+        /// <paramref name="maxPriority"/>, otherwise <see cref="Optional{T}.Empty"/>. Note that
+        /// this method does not return property values that come from inherited or default values.
+        /// </remarks>
+        Optional<T> GetBaseValue<T>(StyledPropertyBase<T> property, BindingPriority maxPriority);
```
Comment (Gillibald) [inline src/Avalonia.Base/IAvaloniaObject.cs:55]: The name doesn't feel natural. `GetValueByPriority` ?

**Comment by Gillibald** [comment]
Is that extra memory consumption coming from the event args?

**Comment by MarchingCube** [comment]
> Is that extra memory consumption coming from the event args?

I think so, but I didn't run memory profiler on it so just a guess.

**Comment by grokys** [comment]
>> Is that extra memory consumption coming from the event args?
> I think so, but I didn't run memory profiler on it so just a guess.

Yeah, quite possibly: we're now doing more property changed notifications. A further improvement will be to pool `AvaloniaPropertyChangedEventArgs<T>` objects.

File: src/Avalonia.Base/IAvaloniaObject.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Base/IAvaloniaObject.cs:55]: The naming comes from WPF: [`GetAnimationBaseValue`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.uielement.getanimationbasevalue?view=netcore-3.1) except that it now takes a binding priority so I removed the "Animation" part from the name.

`GetValueByPriority` might be a good alternative though.

File: src/Avalonia.Controls.DataGrid/DataGrid.cs [hunk 2]
```diff
@@ -767,7 +767,7 @@ private void OnRowDetailsTemplateChanged(AvaloniaPropertyChangedEventArgs e)
         /// <summary>
         /// ItemsProperty property changed handler.
         /// </summary>
-        /// <param name="e">AvaloniaPropertyChangedEventArgs.</param>
+        /// <param name="e">AvaloniaPropertyChangedEventArgsdEventArgs.</param>
```
Comment (MarchingCube) [inline src/Avalonia.Controls.DataGrid/DataGrid.cs:770]: Typo.

**Review by MarchingCube — APPROVED** [review]
I don’t see anything obviously wrong here, time will tell if there are any leftover problems :)

File: src/Avalonia.Base/IAvaloniaObject.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Base/IAvaloniaObject.cs:55]: For the record: we decided to add a `minPriority` parameter to this method and change its name to `GetValueByPriority` in a future PR.

**Pull request merged**